src/main/java/com/egrine/mailspammer/emailrecipient/implementation/InMemoryEmailRecipientService.java

```
@@ -1,23 +1,100 @@
 package com.egrine.mailspammer.emailrecipient.implementation;
 
+import com.egrine.mailspammer.email.DTO.EmailTemplateDTO;
+import com.egrine.mailspammer.email.EmailService;
+import com.egrine.mailspammer.email.EmailTemplate;
+import com.egrine.mailspammer.emailrecipient.DTO.EmailRecipientDTO;
+import com.egrine.mailspammer.emailrecipient.DTO.UpdateEmailRecipientDTO;
 import com.egrine.mailspammer.emailrecipient.EmailRecipient;
 import com.egrine.mailspammer.emailrecipient.EmailRecipientRepository;
 import com.egrine.mailspammer.emailrecipient.EmailRecipientService;
+import com.egrine.mailspammer.user.UserService;
+import com.egrine.mailspammer.utility.Exceptions.RecipientNotFoundException;
+import com.egrine.mailspammer.utility.Exceptions.TemplateNotFoundException;
 import lombok.RequiredArgsConstructor;
 import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.security.core.userdetails.User;
 import org.springframework.stereotype.Service;
 
+import java.util.ArrayList;
 import java.util.Collections;
 import java.util.List;
 
 @Service
 @RequiredArgsConstructor(onConstructor = @__(@Autowired))
 class InMemoryEmailRecipientService implements EmailRecipientService {
     EmailRecipientRepository repository;
+    UserService userService;
+    EmailService emailService;
 
     @Override
     public List<EmailRecipient> getEmailRecipientsById(List<Long> recipientIds) {
         List<EmailRecipient> recipients = repository.getEmailRecipientsById(recipientIds);
         return recipients == null ? Collections.emptyList() : recipients;
     }
+
```

service layer should throw exception if the recipient is not found, you are not using the exception you defined

---

the exception is used in the update function, just like in the templates